aips/aip-103.md

```diff
@@ -0,0 +1,110 @@
+
+This is the advantage of explicit type safety as a completely new validation path would be implemented.  There would be no risk of such proofs being accepted for accounts that require `aud` to be present due to explicit differences in how the proof would be validated gated on the type of public key.
+
+However the drawbacks include:
+- We need to add a new keyless public key type, which may not be needed if we can leverage the existing design.  And avoiding proliferation of keyless public key types is desirable.
+- Requiring imiplementation of a new authentication path in the authenticator, which may be error prone and takes additional engineering effort.
+- Requires more complex changes to the prover as it would need to support a different public inputs hash calculation in order to differentiate between accounts with and without `aud`.  Or it would need to use a different circuit versionentirely.
+
+Thus if we can leverage the existing design, it would be preferable to do so.
+
+## Specification and Implementation Details
+
+This AIP's implementation has three parts -
+
+1. We add an additional private input into the prover.  This value will indicate whether the `aud` check is enabled.  
+- If it is enabled, the circuit will do the status quo set of verifications.  
+- If it is disabled, the circuit will use an empty `aud` value as a private input and skip checking the `aud` field.  A valid JWT still needs to be provided and the public inputs hash calculation will still implicitly include the `aud` field via the `IdCommitment` in the `KeylessPublicKey`, but it will be set to the empty value.  This means that proofs for `aud`-less accounts will be rejected if the account requires `aud` to be present, as the `IdCommitment` will be different.
+
+2. The prover API will also require an update to allow for indiciating whether the `aud` check is enabled.  This will be done by adding a new boolean argument to the `prove` API.
+
+3. The SDK will also need to be updated to support instantiating of such accounts.  This will require adding a new boolean argument to the `KeylessAccount` constructor and constructing the `KeylessPublicKey` and `AccountAddress` appropriately.
+
+## Testing (Optional)
+
+1. Write unit tests for the circuit to verify that it correctly handles the `aud` check.
+2. Write unit tests for the SDK to verify that it correctly instantiates accounts with and without `aud` checks.
+3. Do a manual end-to-end test in devnet/testnet via the SDK once the verification key is updated.
+4. Write smoke tests ensuring that `aud`-less accounts are rejected if the account requires `aud` to be present.
+
+## Security Considerations
+
+The core security considerations are:
+- Making sure the the circuit can securely support `aud`-less accounts.
+- Making sure that such proofs are rejected if the account requires `aud` to be present (as encoded in the `KeylessPublicKey`s `IdCommitment`).
```

Comment:
I am not sure what this line means in saying "proofs are rejected". If the skip aud checks boolean is private, then the verifier will not have access to it. So we cannot accept or reject a proof based on its value

Reply:
I mean the a proof for an audless account is not accepted by an account with the aud set.